File: adventjs/2025/22/solution.js
/**
 * @param {string[][]} maze
 * @returns {boolean}
 */
function canEscape(maze) {
  const rows = maze.length, cols = maze[0].length;
  let startR = -1, startC = -1;
  
  for (let r = 0; r < rows; r++) {
    for (let c = 0; c < cols; c++) {
      if (maze[r][c] === 'S') {
        startR = r;
        startC = c;
        break;
      }
    }
    if (startR !== -1) break;
  }
  
  const queue = [[startR, startC]];
  const visited = new Set([`${startR},${startC}`]);
  const dirs = [[-1, 0], [1, 0], [0, -1], [0, 1]];
  
  while (queue.length) {
    const [r, c] = queue.shift();
    
    if (maze[r][c] === 'E') return true;
    
    for (const [dr, dc] of dirs) {
      const nr = r + dr, nc = c + dc;
      const key = `${nr},${nc}`;
      
      if (nr >= 0 && nr < rows && nc >= 0 && nc < cols && 
          maze[nr][nc] !== '#' && !visited.has(key)) {
        visited.add(key);
        queue.push([nr, nc]);
      }
    }
  }
  
  return false;
}

File: adventjs/2025/22/solution.py
def can_escape(maze: list[list[str]]) -> bool:
  rows, cols = len(maze), len(maze[0])
  start_r, start_c = -1, -1
  
  for r in range(rows):
    for c in range(cols):
      if maze[r][c] == 'S':
        start_r, start_c = r, c
        break
    if start_r != -1:
      break
  
  queue = [(start_r, start_c)]
  visited = {(start_r, start_c)}
  dirs = [(-1, 0), (1, 0), (0, -1), (0, 1)]
  
  while queue:
    r, c = queue.pop(0)
    
    if maze[r][c] == 'E':
      return True
    
    for dr, dc in dirs:
      nr, nc = r + dr, c + dc
      if 0 <= nr < rows and 0 <= nc < cols and maze[nr][nc] != '#' and (nr, nc) not in visited:
        visited.add((nr, nc))
        queue.append((nr, nc))
  
  return False
